File: tools/podcast-transcriber/scripts/sidecar_protocol.py
from __future__ import annotations

import json
import os

READY_PROTOCOL_VERSION = 1


def resolve_sidecar_port(value: str | None = None) -> int:
    raw = os.environ.get("IMMERSIVE_SIDECAR_PORT", "0") if value is None else value
    if not raw.strip().isdigit():
        raise ValueError(f"Invalid sidecar port: {raw}")
    port = int(raw)
    if port < 0 or port > 65535:
        raise ValueError(f"Invalid sidecar port: {raw}")
    return port


def ready_payload(engine: str, pid: int, port: int) -> dict[str, int | str]:
    if not engine or pid <= 0 or port <= 0 or port > 65535:
        raise ValueError("READY payload requires a positive PID and bound port")
    return {
        "engine": engine,
        "protocolVersion": READY_PROTOCOL_VERSION,
        "pid": pid,
        "port": port,
    }


def format_ready_line(engine: str, pid: int, port: int) -> str:
    return json.dumps(ready_payload(engine, pid, port), separators=(",", ":")) + "\n"


def write_ready(engine: str, pid: int, port: int) -> None:
    print(format_ready_line(engine, pid, port), end="", flush=True)

File: tools/zhihu-packer/src/sidecar-protocol.ts
export const READY_PROTOCOL_VERSION = 1;

export interface ReadyPayload {
  engine: string;
  protocolVersion: number;
  pid: number;
  port: number;
}

export function resolveSidecarPort(value = process.env.IMMERSIVE_SIDECAR_PORT ?? "0"): number {
  if (!/^\d+$/.test(value.trim())) {
    throw new Error(`Invalid sidecar port: ${value}`);
  }
  const port = Number(value);
  if (!Number.isInteger(port) || port < 0 || port > 65535) {
    throw new Error(`Invalid sidecar port: ${value}`);
  }
  return port;
}

export function readyPayload(engine: string, pid: number, port: number): ReadyPayload {
  if (!engine || !Number.isInteger(pid) || pid <= 0 || !Number.isInteger(port) || port <= 0 || port > 65535) {
    throw new Error("READY payload requires a positive PID and bound port");
  }
  return {
    engine,
    protocolVersion: READY_PROTOCOL_VERSION,
    pid,
    port,
  };
}

export function formatReadyLine(engine: string, pid: number, port: number): string {
  return `${JSON.stringify(readyPayload(engine, pid, port))}\n`;
}

export function writeReady(engine: string, pid: number, port: number): void {
  process.stdout.write(formatReadyLine(engine, pid, port));
}
